network/wsNetwork.go

```diff
@@ -2370,3 +2383,48 @@ func (wn *WebsocketNetwork) postMessagesOfInterestThread() {
 func (wn *WebsocketNetwork) SubstituteGenesisID(rawURL string) string {
 	return strings.Replace(rawURL, "{genesisID}", wn.GenesisID, -1)
 }
+
+func (wn *WebsocketNetwork) makePeerSendable(wp *wsPeer) {
+	wn.sendSchedulerMu.Lock()
+	defer wn.sendSchedulerMu.Unlock()
+	didAdd := wn.sendScheduler.add(wp)
+	if didAdd {
+		wn.sendSchedulerCond.Broadcast()
+	}
+}
+
+func (wn *WebsocketNetwork) sendSchedulerThread() {
+	defer wn.wg.Done()
+	var wp *wsPeer
+	for {
+		wn.sendSchedulerMu.Lock()
+		if wp != nil {
+			wp.inSendScheduler = false
+			// previous round's peer
+			if wp.hasWritesPending() {
+				wn.sendScheduler.add(wp)
+			}
+			wp = nil
+		}
+		for wp == nil {
+			select {
+			case <-wn.ctx.Done():
+				wn.sendSchedulerMu.Unlock()
+				return
+			default:
+			}
+			wp = wn.sendScheduler.next()
```

Comment:
This is `pop()`ing the peer from the scheduler under the hood. 
What if the peer has more than 1 message pending?
The calls to `makePeerSendable()` are done when a new item is queued, that means the peer will stall until next item is queued to send the previous one.

Reply:
after a peer has sent the send thread checks to see if there's anything in the peer's send queues; if yes, then back in the scheduler heap; if not then clear the inScheduler bool and drop it and wait for the next makePeerSendable().
_This is not perfectly thread safe_. There's a race window where the peer adds to its queue after the scheduler's queue check and before the scheduler makes the peer re-schedulable. This is an expedient 'usually good enough' to see if an outbound scheduler is useful to our problems.